notebooks/utils/estimation.py

```diff
@@ -0,0 +1,229 @@
+        For a complete list of estimator, see `utils.covariance`.
+
+    """
+
+    def __init__(self, rate: float = 256, flickering_frequencies: tuple = (13, 17, 21),
+                 estimator: str = 'scm', **kwargs):
+        """Init."""
+        self._estimator = estimator
+        self._design_filters(rate, flickering_frequencies, **kwargs)
+
+    def fit(self, X: np.array, y: np.array = None) -> object:
+        """Fit.
+
+        Do nothing. For compatibility purpose.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_trials, n_channels, n_samples)
+            ndarray of trials.
+        y : ndarray shape (n_trials,)
+            labels corresponding to each trial, not used.
+
+        Returns
+        -------
+        self : SSVEPCovariance instance
+            The SSVEPCovariance instance.
+        """
+        return self
+
+    def transform(self, X: np.array) -> np.array:
+        """Estimate covariance matrices.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_trials, n_channels, n_samples)
+            ndarray of trials.
+
+        Returns
+        -------
+        covmats : ndarray, shape (n_trials, n_channels * n_frequencies, n_channels * n_frequencies)
+            SSVEP covariance matrices for each trials.
+        """
+        filtered = []
+        for filter in self.filters:
+            filtered_X = filter(X, axis=-1)
+            filtered.append(filtered_X)
+        mega_X = np.concatenate(filtered, axis=1)
+        out = covariances(mega_X, estimator=self._estimator)
+        return out
+
+    def _design_filters(self, fs: float, flickering_frequencies: tuple,
+                        bandwidth: float = 4, order: int = 2,
+                        filter_design: str = 'iir', padlen: int = None):
+        self.filters = []
+        for flickering_frequency in flickering_frequencies:
+            frequencies = [flickering_frequency - bandwidth / 2,
+                           flickering_frequency + bandwidth / 2]  # TODO/Question : shall we consider harmonics as well ?
+            filter_args = design_filter(frequencies=frequencies, order=order, fs=fs, filter_design=filter_design)
+            filter_func = functools.partial(filtfilt, *filter_args, padlen=padlen)
+            self.filters.append(filter_func)
+
+
+class SSVEPPotato(BaseEstimator, TransformerMixin):
+    """Estimate potato field per class
+        Parameters
+        ----------
+        classes: classes to consider in potato field (one potato per class)
+        potato_threshold: threshold for each potato
+        field_threshold: threshold for the potato field
+        n_iter_max: max iteration allowed for potato convergence
+        pos_label: label for accepted trials in predict
+        neg_label: label for rejected trials in predict
+    """
+
+    def __init__(self, classes: tuple = None, potato_threshold: float = .1, field_threshold: float = .1,
```

Comment:
0.1 is a very high value, giving a risk of many false negatives. field_threshold should be fixed to 0.01.